clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -63,91 +64,90 @@ public void configure(Map<String, ?> configs) {}
     @Override
     public void init(List<KafkaMetric> metrics) {
         synchronized (LOCK) {
-            for (KafkaMetric metric : metrics)
-                addAttribute(metric);
-            for (KafkaMbean mbean : mbeans.values())
+            for (KafkaMetric metric : metrics) {
+                final KafkaMbean mbean = addAttribute(metric);
+                if (!this.mbeans.containsKey(mbean.name())) {
+                    this.mbeans.put(mbean.name(), mbean);
+                }
+            }
+            for (KafkaMbean mbean : this.mbeans.values()) {
                 reregister(mbean);
+            }
         }
     }
 
     @Override
     public void metricChange(KafkaMetric metric) {
         synchronized (LOCK) {
-            KafkaMbean mbean = addAttribute(metric);
+            final KafkaMbean mbean = addAttribute(metric);
+            if (!this.mbeans.containsKey(mbean.name())) {
+                this.mbeans.put(mbean.name(), mbean);
+            }
             reregister(mbean);
         }
     }
 
     @Override
     public void metricRemoval(KafkaMetric metric) {
         synchronized (LOCK) {
-            KafkaMbean mbean = removeAttribute(metric);
-            if (mbean != null) {
-                if (mbean.metrics.isEmpty())
-                    unregister(mbean);
-                else
-                    reregister(mbean);
+            final KafkaMbean mbean = removeAttribute(metric);
+            if (mbean.isEmpty()) {
+                unregister(mbean);
+                this.mbeans.remove(mbean.name());
+            } else {
+                reregister(mbean);
             }
         }
     }
 
     private KafkaMbean removeAttribute(KafkaMetric metric) {
-        MetricName metricName = metric.metricName();
-        String mBeanName = getMBeanName(metricName);
-        KafkaMbean mbean = this.mbeans.get(mBeanName);
-        if (mbean != null)
+        final MetricName metricName = metric.metricName();
+        final KafkaMbean mbean = getMBean(metricName);
+        if (!mbean.isEmpty()) {
             mbean.removeAttribute(metricName.name());
+        }
         return mbean;
     }
 
     private KafkaMbean addAttribute(KafkaMetric metric) {
-        try {
-            MetricName metricName = metric.metricName();
-            String mBeanName = getMBeanName(metricName);
-            if (!this.mbeans.containsKey(mBeanName))
-                mbeans.put(mBeanName, new KafkaMbean(mBeanName));
```

Comment:
While updating unit tests (as requested by @guozhangwang) and reading this code I was puzzled much with the fact that non-thread-safe `HashMap` is modified in this method. Then I've found that the method is called only from a synchronized block, so it's safe. But still to improve readability and help future maintainers I've decided to move all such modifications to the appropriate synchronized blocks and keep helper methods just for "routine".